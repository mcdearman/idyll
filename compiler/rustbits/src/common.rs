use std::ops::{Index, Range};

#[derive(Debug, Clone, PartialEq)]
pub struct Loc {
    pub start: u32,
    pub end: u32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Located<T> {
    pub value: T,
    pub loc: Loc,
}

impl From<Loc> for Range<usize> {
    fn from(loc: Loc) -> Self {
        loc.start as usize..loc.end as usize
    }
}

impl From<Range<usize>> for Loc {
    fn from(range: Range<usize>) -> Self {
        Self {
            start: range.start as u32,
            end: range.end as u32,
        }
    }
}

impl Index<Loc> for str {
    type Output = str;

    fn index(&self, index: Loc) -> &Self::Output {
        // Convert the Span to a Range<usize>
        let range: Range<usize> = index.into();

        // Ensure the range is within the bounds of the string
        if range.start > self.len() || range.end > self.len() {
            panic!("Index out of bounds");
        }

        // Return the slice of the string for the given range
        &self[range]
    }
}
