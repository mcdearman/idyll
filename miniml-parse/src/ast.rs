#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    decls: Vec<Decl>,
    span: Span,
}

impl Root {
    pub fn new(decls: Vec<Decl>, span: Span) -> Self {
        Self { decls, span }
    }

    pub fn decls(&self) -> &[Decl] {
        &self.decls
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum DeclKind {
    DataType(DataType),
    Let { pattern: Pattern, expr: Expr },
}

#[derive(Debug, Clone, PartialEq)]
pub struct DataType {
    name: Ident,
    kind: Box<DataTypeKind>,
    span: Span,
}

impl DataType {
    pub fn new(name: Ident, kind: DataTypeKind, span: Span) -> Self {
        Self {
            name,
            kind: Box::new(kind),
            span,
        }
    }

    pub fn name(&self) -> &Ident {
        &self.name
    }

    pub fn kind(&self) -> &DataTypeKind {
        &self.kind
    }

    pub fn span(&self) -> &Span {
        &self.span
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum DataTypeKind {
    Record {
        fields: Vec<(Ident, TypeHint)>,
    },
    Sum {
        variants: Vec<(Ident, Option<TypeHint>)>,
    },
    Product {
        fields: Vec<TypeHint>,
    },
}
