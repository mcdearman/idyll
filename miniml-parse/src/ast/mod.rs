pub mod binary_op;
pub mod binary_op_kind;
pub mod decl;
pub mod decl_kind;
pub mod expr;
pub mod expr_kind;
pub mod ident;
pub mod let_arm;
pub mod lit;
pub mod pattern;
pub mod pattern_kind;
pub mod root;
pub mod unary_op;
pub mod unary_op_kind;
