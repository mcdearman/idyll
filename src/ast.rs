use super::node::Node;
use crate::{num::Num, token::Token};

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    pub classes: Vec<Node<Class>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Class {
    pub name: Node<String>,
    pub fields: Vec<Node<Field>>,
    pub methods: Vec<Node<Method>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    pub name: Node<String>,
    pub expr: Node<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Method {
    pub name: Node<String>,
    pub params: Vec<Node<Pattern>>,
    pub body: Node<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit(Lit),
    Ident(String),
    Lambda {
        params: Vec<Node<Pattern>>,
        body: Node<Self>,
    },
    Apply {
        fun: Node<Self>,
        args: Vec<Node<Self>>,
    },
    Let {
        pat: Node<Pattern>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Fn {
        name: Node<String>,
        params: Vec<Node<Pattern>>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Match {
        expr: Node<Self>,
        cases: Vec<Node<MatchCase>>,
    },
    If {
        cond: Node<Self>,
        then: Node<Self>,
        else_: Node<Self>,
    },
    Prefix {
        op: Node<PrefixOp>,
        expr: Node<Self>,
    },
    Infix {
        op: Node<InfixOp>,
        lhs: Node<Self>,
        rhs: Node<Self>,
    },
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchCase {
    pub pattern: Node<Pattern>,
    pub expr: Node<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Num(Num),
    Bool(bool),
    String(String),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Pattern {
    Lit(Lit),
    Ident(String),
    Wildcard,
    Unit,
    // Tuple(Vec<Node<Self>>),
    // Pair {
    //     head: Node<Self>,
    //     tail: Node<Self>,
    // },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PrefixOp {
    Neg,
    Not,
}

impl ToString for PrefixOp {
    fn to_string(&self) -> String {
        match self {
            Self::Neg => "-",
            Self::Not => "!",
        }
        .to_string()
    }
}

impl From<Token> for PrefixOp {
    fn from(token: Token) -> Self {
        match token {
            Token::Minus => Self::Neg,
            Token::Bang => Self::Not,
            kind => panic!("Not an prefix operator: {:?}", kind),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InfixOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    // Pow,
    Eq,
    Neq,
    Lt,
    Gt,
    Leq,
    Geq,
    // And,
    // Or,
}

impl ToString for InfixOp {
    fn to_string(&self) -> String {
        match self {
            Self::Add => "+",
            Self::Sub => "-",
            Self::Mul => "*",
            Self::Div => "/",
            Self::Mod => "%",
            // Self::Pow => "^",
            Self::Eq => "==",
            Self::Neq => "!=",
            Self::Lt => "<",
            Self::Gt => ">",
            Self::Leq => "<=",
            Self::Geq => ">=",
            // Self::And => "&&",
            // Self::Or => "||",
        }
        .to_string()
    }
}

impl From<Token> for InfixOp {
    fn from(token: Token) -> Self {
        match token {
            Token::Plus => Self::Add,
            Token::Minus => Self::Sub,
            Token::Star => Self::Mul,
            Token::Slash => Self::Div,
            Token::Percent => Self::Mod,
            // Token::Caret => Self::Pow,
            Token::Eq => Self::Eq,
            Token::Neq => Self::Neq,
            Token::Lt => Self::Lt,
            Token::Gt => Self::Gt,
            Token::Leq => Self::Leq,
            Token::Geq => Self::Geq,
            // Token::And => Self::And,
            // Token::Or => Self::Or,
            kind => panic!("Not an infix operator: {:?}", kind),
        }
    }
}
