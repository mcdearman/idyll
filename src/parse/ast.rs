use num_rational::Rational64;

use crate::{
    lex::token::Token,
    utils::{ident::Ident, intern::InternedString, span::Span},
};

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    pub decls: Vec<Decl>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Decl {
    pub kind: DeclKind,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq)]
pub enum DeclKind {
    Def(Pattern, Expr),
    Fn(Ident, Vec<Pattern>, Expr),
    FnMatch(Vec<(Ident, Vec<Pattern>, Expr)>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Expr {
    pub kind: Box<ExprKind>,
    pub span: Span,
}

impl Expr {
    pub fn new(kind: ExprKind, span: Span) -> Self {
        Self {
            kind: Box::new(kind),
            span,
        }
    }

    pub fn is_lambda(&self) -> bool {
        matches!(self.kind.as_ref(), &ExprKind::Lambda(_, _))
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind {
    Lit(Lit),
    Var(Ident),
    Apply(Expr, Vec<Expr>),
    Lambda(Vec<Pattern>, Expr),
    UnaryOp(UnaryOp, Expr),
    BinaryOp(BinaryOp, Expr, Expr),
    Or(Expr, Expr),
    And(Expr, Expr),
    Let(Pattern, Expr, Expr),
    Fn(Ident, Vec<Pattern>, Expr, Expr),
    If(Expr, Expr, Expr),
    Match(Expr, Vec<(Pattern, Expr)>),
    List(Vec<Expr>),
    Unit,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct UnaryOp {
    pub kind: UnaryOpKind,
    pub span: Span,
}

impl From<UnaryOp> for InternedString {
    fn from(op: UnaryOp) -> Self {
        InternedString::from(op.kind.to_string())
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum UnaryOpKind {
    Neg,
    Not,
}

impl From<Token> for UnaryOpKind {
    fn from(t: Token) -> Self {
        match t {
            Token::Minus => Self::Neg,
            Token::Not => Self::Not,
            _ => unreachable!(),
        }
    }
}

impl ToString for UnaryOpKind {
    fn to_string(&self) -> String {
        match self {
            Self::Neg => "__neg__".to_string(),
            Self::Not => "__not__".to_string(),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BinaryOp {
    pub kind: BinaryOpKind,
    pub span: Span,
}

impl From<BinaryOp> for InternedString {
    fn from(op: BinaryOp) -> Self {
        InternedString::from(op.kind.to_string())
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum BinaryOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Rem,
    Pow,
    Eq,
    Neq,
    Lt,
    Lte,
    Gt,
    Gte,
    Pair,
}

impl From<Token> for BinaryOpKind {
    fn from(t: Token) -> Self {
        match t {
            Token::Plus => Self::Add,
            Token::Minus => Self::Sub,
            Token::Star => Self::Mul,
            Token::Slash => Self::Div,
            Token::Percent => Self::Rem,
            Token::Caret => Self::Pow,
            Token::Eq => Self::Eq,
            Token::Neq => Self::Neq,
            Token::Lt => Self::Lt,
            Token::Leq => Self::Lte,
            Token::Gt => Self::Gt,
            Token::Geq => Self::Gte,
            Token::DoubleColon => Self::Pair,
            _ => unreachable!(),
        }
    }
}

impl ToString for BinaryOpKind {
    fn to_string(&self) -> String {
        match self {
            Self::Add => "__add__".to_string(),
            Self::Sub => "__sub__".to_string(),
            Self::Mul => "__mul__".to_string(),
            Self::Div => "__div__".to_string(),
            Self::Rem => "__rem__".to_string(),
            Self::Pow => "__pow__".to_string(),
            Self::Eq => "__eq__".to_string(),
            Self::Neq => "__neq__".to_string(),
            Self::Lt => "__lt__".to_string(),
            Self::Lte => "__lte__".to_string(),
            Self::Gt => "__gt__".to_string(),
            Self::Gte => "__gte__".to_string(),
            Self::Pair => "__pair__".to_string(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Pattern {
    pub kind: Box<PatternKind>,
    pub span: Span,
}

impl Pattern {
    pub fn new(kind: PatternKind, span: Span) -> Self {
        Self {
            kind: Box::new(kind),
            span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum PatternKind {
    Wildcard,
    Lit(Lit),
    Ident(Ident, Option<TypeHint>),
    List(Vec<Pattern>),
    Pair(Pattern, Pattern),
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeHint {
    pub kind: Box<TypeHintKind>,
    pub span: Span,
}

impl TypeHint {
    pub fn new(kind: TypeHintKind, span: Span) -> Self {
        Self {
            kind: Box::new(kind),
            span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeHintKind {
    Byte,
    Int,
    Rational,
    Real,
    Bool,
    String,
    Char,
    Ident(Ident),
    Fn(Vec<TypeHint>, TypeHint),
    List(TypeHint),
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Byte(u8),
    Int(i64),
    Rational(Rational64),
    Real(f64),
    Bool(bool),
    String(InternedString),
    Char(char),
}
