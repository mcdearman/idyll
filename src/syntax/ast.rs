use super::token::Token;
use crate::util::{intern::InternedString, node::Node};
use num_rational::Rational64;

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    pub items: Vec<Node<Item>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Item {
    Def {
        name: Node<InternedString>,
        expr: Node<Expr>,
    },
    Fn {
        name: Node<InternedString>,
        params: Vec<Node<InternedString>>,
        body: Node<Expr>,
    },
    Expr(Expr),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit(Lit),
    Ident(InternedString),
    Lambda {
        params: Vec<Node<InternedString>>,
        body: Node<Self>,
    },
    Apply {
        fun: Node<Self>,
        args: Vec<Node<Self>>,
    },
    Let {
        pat: Node<Pattern>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Fn {
        name: Node<InternedString>,
        params: Vec<Node<Pattern>>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Match {
        expr: Node<Self>,
        cases: Vec<Node<MatchCase>>,
    },
    If {
        cond: Node<Self>,
        then: Node<Self>,
        else_: Node<Self>,
    },
    Prefix {
        op: Node<PrefixOp>,
        expr: Node<Self>,
    },
    Infix {
        op: Node<InfixOp>,
        lhs: Node<Self>,
        rhs: Node<Self>,
    },
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchCase {
    pub pattern: Node<Pattern>,
    pub expr: Node<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Num(Rational64),
    Bool(bool),
    String(InternedString),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Pattern {
    Lit(Lit),
    Ident(InternedString),
    Wildcard,
    Unit,
    // Tuple(Vec<Node<Self>>),
    // Pair {
    //     head: Node<Self>,
    //     tail: Node<Self>,
    // },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PrefixOp {
    Neg,
    Not,
}

impl ToString for PrefixOp {
    fn to_string(&self) -> String {
        match self {
            Self::Neg => "-",
            Self::Not => "!",
        }
        .to_string()
    }
}

impl From<Token> for PrefixOp {
    fn from(token: Token) -> Self {
        match token {
            Token::Minus => Self::Neg,
            Token::Bang => Self::Not,
            kind => panic!("Not an prefix operator: {:?}", kind),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InfixOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    // Pow,
    Eq,
    Neq,
    Lt,
    Gt,
    Leq,
    Geq,
    // And,
    // Or,
}

impl ToString for InfixOp {
    fn to_string(&self) -> String {
        match self {
            Self::Add => "+",
            Self::Sub => "-",
            Self::Mul => "*",
            Self::Div => "/",
            Self::Mod => "%",
            // Self::Pow => "^",
            Self::Eq => "==",
            Self::Neq => "!=",
            Self::Lt => "<",
            Self::Gt => ">",
            Self::Leq => "<=",
            Self::Geq => ">=",
            // Self::And => "&&",
            // Self::Or => "||",
        }
        .to_string()
    }
}

impl From<Token> for InfixOp {
    fn from(token: Token) -> Self {
        match token {
            Token::Plus => Self::Add,
            Token::Minus => Self::Sub,
            Token::Star => Self::Mul,
            Token::Slash => Self::Div,
            Token::Percent => Self::Mod,
            // Token::Caret => Self::Pow,
            Token::Eq => Self::Eq,
            Token::Neq => Self::Neq,
            Token::Lt => Self::Lt,
            Token::Gt => Self::Gt,
            Token::Leq => Self::Leq,
            Token::Geq => Self::Geq,
            // Token::And => Self::And,
            // Token::Or => Self::Or,
            kind => panic!("Not an infix operator: {:?}", kind),
        }
    }
}
