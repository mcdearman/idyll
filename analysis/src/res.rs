/*
 * This module resolves names in the AST and produces an IR similar
 * to the AST but with all names resolved to their unique IDs. Names
 * that shadow names from an outer scope are given a new unique ID.
 */
use common::{intern::InternedString, node::Node, span::Span, unique_id::UniqueId};
use log::trace;
use num_rational::Rational64;
use std::{cell::RefCell, collections::HashMap, rc::Rc};
use syntax::ast;

#[derive(Debug, Clone, PartialEq)]
pub struct ResError {
    pub msg: InternedString,
    pub span: Span,
}

pub type ResResult<T> = Result<T, ResError>;

#[derive(Debug, Clone, PartialEq)]
pub struct Env {
    parent: Option<Rc<RefCell<Env>>>,
    data: HashMap<InternedString, UniqueId>,
}

impl Env {
    pub fn new() -> Rc<RefCell<Self>> {
        Rc::new(RefCell::new(Self {
            parent: None,
            data: HashMap::new(),
        }))
    }

    pub fn new_with_parent(parent: Rc<RefCell<Self>>) -> Rc<RefCell<Self>> {
        Rc::new(RefCell::new(Self {
            parent: Some(parent),
            data: HashMap::new(),
        }))
    }

    pub fn find(&self, name: &InternedString) -> Option<UniqueId> {
        if let Some(id) = self.data.get(name) {
            Some(*id)
        } else if let Some(parent) = &self.parent {
            parent.borrow().find(name)
        } else {
            None
        }
    }

    pub fn find_in_scope(&self, name: &InternedString) -> Option<UniqueId> {
        self.data.get(name).copied()
    }

    pub fn find_in_parent(&self, name: &InternedString) -> Option<UniqueId> {
        if let Some(parent) = &self.parent {
            parent.borrow().find(name)
        } else {
            None
        }
    }

    pub fn define(&mut self, name: InternedString) -> UniqueId {
        let id = UniqueId::gen();
        self.data.insert(name, id);
        id
    }

    pub fn insert(&mut self, name: InternedString, id: UniqueId) {
        self.data.insert(name, id);
    }

    pub fn define_if_absent(&mut self, name: InternedString) -> UniqueId {
        if let Some(id) = self.data.get(&name) {
            *id
        } else {
            let id = UniqueId::gen();
            self.data.insert(name, id);
            id
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Root {
    pub items: Vec<Node<Item>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Item {
    Def {
        pat: Node<Pattern>,
        expr: Node<Expr>,
    },
    Fn {
        name: Node<UniqueId>,
        params: Vec<Node<Pattern>>,
        body: Node<Expr>,
    },
    Expr(Expr),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit(Node<Lit>),
    Ident(Node<UniqueId>),
    Lambda {
        params: Vec<Node<Pattern>>,
        body: Node<Self>,
    },
    Apply {
        fun: Node<Self>,
        args: Vec<Node<Self>>,
    },
    Let {
        pat: Node<Pattern>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Fn {
        name: Node<UniqueId>,
        params: Vec<Node<Pattern>>,
        expr: Node<Self>,
        body: Node<Self>,
    },
    Match {
        expr: Node<Self>,
        cases: Vec<Node<MatchCase>>,
    },
    If {
        cond: Node<Self>,
        then: Node<Self>,
        else_: Node<Self>,
    },
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MatchCase {
    pub pattern: Node<Pattern>,
    pub expr: Node<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Num(Rational64),
    Bool(bool),
    String(InternedString),
}

impl From<ast::Lit> for Lit {
    fn from(lit: ast::Lit) -> Self {
        match lit {
            ast::Lit::Num(num) => Self::Num(num),
            ast::Lit::Bool(b) => Self::Bool(b),
            ast::Lit::String(s) => Self::String(s),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Pattern {
    Lit(Lit),
    Ident(UniqueId),
    Wildcard,
    Unit,
}

pub fn resolve(
    env: Rc<RefCell<Env>>,
    root: &Node<ast::Root>,
) -> (Option<Node<Root>>, Vec<ResError>) {
    let mut errors = vec![];
    let mut items = vec![];
    for item in root.inner().clone().items {
        match resolve_item(env.clone(), item) {
            Ok(i) => {
                // trace!("env: {:#?}", env.borrow());
                items.push(i);
            }
            Err(err) => {
                trace!("env: {:#?}", env.borrow());
                errors.push(err)
            }
        }
    }
    if items.is_empty() {
        (None, errors)
    } else {
        (Some(Node::new(Root { items }, root.span())), errors)
    }
}

fn resolve_item(env: Rc<RefCell<Env>>, item: Node<ast::Item>) -> ResResult<Node<Item>> {
    trace!("item env: {:#?}", env.borrow());
    match &*item {
        ast::Item::Def { pat, expr } => {
            // let def_env = Env::new_with_parent(env.clone());
            let expr = resolve_expr(env.clone(), expr)?;
            let pat = resolve_pattern(env.clone(), pat)?;
            trace!("def env: {:#?}", env.borrow());
            Ok(Node::new(Item::Def { pat, expr }, item.span()))
        }
        ast::Item::Fn { name, params, body } => {
            let name = Node::new(env.borrow_mut().define(name.inner().clone()), name.span());
            let mut new_params = vec![];
            let fn_env = Env::new_with_parent(env.clone());
            for p in params {
                let pat = resolve_pattern(fn_env.clone(), &p)?;
                new_params.push(pat);
            }
            let body = resolve_expr(fn_env.clone(), &body)?;
            Ok(Node::new(
                Item::Fn {
                    name,
                    params: new_params,
                    body,
                },
                item.span(),
            ))
        }
        ast::Item::Expr(expr) => {
            let expr = resolve_expr(env.clone(), &Node::new(expr.clone(), item.span()))?;
            Ok(Node::new(Item::Expr(expr.inner().clone()), item.span()))
        }
    }
}

fn resolve_expr(env: Rc<RefCell<Env>>, expr: &Node<ast::Expr>) -> ResResult<Node<Expr>> {
    match expr.inner() {
        ast::Expr::Ident(ident) => {
            if let Some(name) = env.borrow().find(ident) {
                Ok(Node::new(
                    Expr::Ident(Node::new(name, expr.span())),
                    expr.span(),
                ))
            } else {
                Err(ResError {
                    msg: InternedString::from(&*format!(
                        "name '{:?}' is not defined",
                        expr.inner()
                    )),
                    span: expr.span(),
                })
            }
        }
        ast::Expr::Lit(l) => Ok(Node::new(
            Expr::Lit(Node::new(l.clone().into(), expr.span())),
            expr.span(),
        )),
        ast::Expr::Let { pat, expr, body } => {
            let let_env = Env::new_with_parent(env.clone());
            let expr = resolve_expr(let_env.clone(), &expr.clone())?;
            let pat = resolve_pattern(let_env.clone(), pat)?;
            trace!("let env: {:#?}", let_env.borrow());
            let body = resolve_expr(let_env.clone(), body)?;
            Ok(Node::new(
                Expr::Let {
                    pat,
                    expr: expr.clone(),
                    body,
                },
                expr.span(),
            ))
        }
        ast::Expr::Fn {
            name,
            params,
            expr,
            body,
        } => {
            let name = Node::new(env.borrow_mut().define(name.inner().clone()), name.span());
            let mut new_params = vec![];
            let fn_env = Env::new_with_parent(env.clone());
            for p in params {
                let pat = resolve_pattern(fn_env.clone(), p)?;
                new_params.push(pat);
            }
            let expr = resolve_expr(fn_env.clone(), expr)?;
            let body = resolve_expr(fn_env, body)?;
            Ok(Node::new(
                Expr::Fn {
                    name,
                    params: new_params,
                    expr: expr.clone(),
                    body,
                },
                expr.span(),
            ))
        }
        ast::Expr::Apply { fun, args } => {
            let fun = resolve_expr(env.clone(), fun)?;
            let args = args
                .iter()
                .map(|arg| resolve_expr(env.clone(), arg))
                .collect::<ResResult<Vec<_>>>()?;
            Ok(Node::new(
                Expr::Apply {
                    fun: fun.clone(),
                    args,
                },
                expr.span(),
            ))
        }
        ast::Expr::Lambda { params, body } => {
            let mut new_params = vec![];
            let lambda_env = Env::new_with_parent(env.clone());
            for p in params {
                let pat = resolve_pattern(lambda_env.clone(), p)?;
                new_params.push(pat);
            }
            let body = resolve_expr(lambda_env, body)?;
            Ok(Node::new(
                Expr::Lambda {
                    params: new_params,
                    body,
                },
                expr.span(),
            ))
        }
        ast::Expr::Match { expr, cases } => {
            let expr = resolve_expr(env.clone(), expr)?;
            let cases = cases
                .iter()
                .map(|case| {
                    let pattern = resolve_pattern(env.clone(), &case.pattern)?;
                    let expr = resolve_expr(env.clone(), &case.expr)?;
                    Ok(Node::new(
                        MatchCase {
                            pattern: pattern.clone(),
                            expr: expr.clone(),
                        },
                        case.span(),
                    ))
                })
                .collect::<ResResult<Vec<_>>>()?;
            Ok(Node::new(
                Expr::Match {
                    expr: expr.clone(),
                    cases,
                },
                expr.span(),
            ))
        }
        ast::Expr::If { cond, then, else_ } => {
            let cond = resolve_expr(env.clone(), cond)?;
            let then = resolve_expr(env.clone(), then)?;
            let else_ = resolve_expr(env, else_)?;
            Ok(Node::new(
                Expr::If {
                    cond: cond.clone(),
                    then: then.clone(),
                    else_: else_.clone(),
                },
                expr.span(),
            ))
        }
        ast::Expr::Prefix { op, expr } => {
            let op_id = Node::new(
                env.borrow()
                    .find(&InternedString::from(op.to_string()))
                    .unwrap(),
                op.span(),
            );
            let expr = resolve_expr(env, expr)?;
            Ok(Node::new(
                Expr::Apply {
                    fun: Node::new(Expr::Ident(op_id), op.span()),
                    args: vec![expr.clone()],
                },
                expr.span(),
            ))
        }
        ast::Expr::Infix { op, lhs, rhs } => {
            let op_id = Node::new(
                env.borrow()
                    .find(&InternedString::from(op.to_string()))
                    .unwrap(),
                op.span(),
            );
            let lhs = resolve_expr(env.clone(), lhs)?;
            let rhs = resolve_expr(env, rhs)?;
            Ok(Node::new(
                Expr::Apply {
                    fun: Node::new(Expr::Ident(op_id), op.span()),
                    args: vec![lhs.clone(), rhs.clone()],
                },
                expr.span(),
            ))
        }
        ast::Expr::Unit => Ok(Node::new(Expr::Unit, expr.span())),
    }
}

fn resolve_pattern(
    env: Rc<RefCell<Env>>,
    pattern: &Node<ast::Pattern>,
    // rec: bool,
    // def: bool,
) -> ResResult<Node<Pattern>> {
    match pattern.inner() {
        ast::Pattern::Lit(l) => Ok(Node::new(Pattern::Lit(l.clone().into()), pattern.span())),
        ast::Pattern::Ident(ident) => {
            // if def {
            // if let Some(name) = env.borrow().find(ident) {
            //     return Err(ResError {
            //         msg: InternedString::from(&*format!(
            //             "name '{:?}' is already defined as {:?}",
            //             pattern.inner(),
            //             name
            //         )),
            //         span: pattern.span(),
            //     });
            // }
            let name = env.borrow_mut().define(ident.clone());
            Ok(Node::new(Pattern::Ident(name), pattern.span()))
            // } else {
            //     if rec {
            //         if let Some(name) = env.borrow().find(ident) {
            //             Ok(Node::new(Pattern::Ident(name), pattern.span()))
            //         } else {
            //             Err(ResError {
            //                 msg: InternedString::from(&*format!(
            //                     "name '{:?}' is not defined",
            //                     pattern.inner()
            //                 )),
            //                 span: pattern.span(),
            //             })
            //         }
            //     } else {
            //         if let Some(name) = env.borrow().find_in_scope(ident) {
            //             Ok(Node::new(Pattern::Ident(name), pattern.span()))
            //         } else {
            //             Err(ResError {
            //                 msg: InternedString::from(&*format!(
            //                     "name '{:?}' is not defined",
            //                     pattern.inner()
            //                 )),
            //                 span: pattern.span(),
            //             })
            //         }
            //     }
            // }
        }
        ast::Pattern::Wildcard => Ok(Node::new(Pattern::Wildcard, pattern.span())),
        ast::Pattern::Unit => Ok(Node::new(Pattern::Unit, pattern.span())),
    }
}

mod tests {
    // #[test]
    // fn res_nested() {
    //     let (ast, errors) = crate::syntax::parse::parse("let x = 1 in let x = 2 in x + 1");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }
    // #[test]
    // fn res_def() {
    //     let (ast, errors) = crate::syntax::parse::parse("x = 1");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }

    // #[test]
    // fn res_def_error() {
    //     let (ast, errors) = crate::syntax::parse::parse("x = x");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (_, errors) = super::resolve(&ast.unwrap());
    //     assert!(!errors.is_empty());
    // }

    // #[test]
    // fn res_fn_def() {
    //     let (ast, errors) = crate::syntax::parse::parse("add x y = x + y");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }

    // #[test]
    // fn res_rec_fn_def() {
    //     let (ast, errors) = crate::syntax::parse::parse("f x = f x");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }

    // #[test]
    // fn res_let() {
    //     let (ast, errors) = crate::syntax::parse::parse("let x = 1 in x");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }

    // #[test]
    // fn res_rec_let() {
    //     let (ast, errors) = crate::syntax::parse::parse("let f x = f x in f 1");
    //     if !errors.is_empty() {
    //         panic!("parse error: {:?}", errors);
    //     }
    //     let (res, errors) = super::resolve(&ast.unwrap());
    //     if !errors.is_empty() {
    //         panic!("resolve error: {:?}", errors);
    //     }
    //     insta::assert_debug_snapshot!(res);
    // }
}
