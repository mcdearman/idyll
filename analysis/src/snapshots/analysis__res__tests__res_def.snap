---
source: analysis/src/res.rs
expression: "test_helper(\"x = 1\")"
---
Root {
    items: [
        Def {
            pat: Ident(
                UniqueId(2),
            ) @ 0..1,
            expr: Lit(
                Num(
                    Ratio {
                        numer: 1,
                        denom: 1,
                    },
                ) @ 4..5,
            ) @ 4..5,
        } @ 0..5,
    ],
} @ 0..5
